aten/src/ATen/core/type.cpp

```diff
@@ -1189,11 +1190,16 @@ c10::optional<TypePtr> UnionType::subtractTypeSet(std::vector<TypePtr>& to_subtr
 
 OptionalType::OptionalType(TypePtr contained)
                            : UnionType({contained, NoneType::get()}, TypeKind::OptionalType) {
+  bool is_numbertype = false;
+  if (auto as_union = contained->cast<UnionType>()) {
+    is_numbertype = as_union->containedTypes().size() == 3 &&
+                    as_union->canHoldType(NumberType::get());
```

Comment:
eventually it would be nice to replace `canHoldType` with `subtypeOf` and remove `canHoldType`, since it's a duplicate API of the more general one